lib/dry/validation/schema.rb

```diff
@@ -92,6 +94,14 @@ def [](name)
       def predicates
         self.class.predicates
       end
+
+      private
+
+      def validate_input(input)
+        unless input.is_a?(Hash)
+          raise ArgumentError, "expected a Hash but received +#{input.inspect}+"
+        end
```

Comment:
I  am also faced with this problem and think that we shouldn't raise error in case if our subject doesn't respond to appropriate attr or `[]` method. `Rule::Key` automatically adds `key?` predicate which can prevent such sort of errors. 
https://github.com/dryrb/dry-validation/blob/master/lib/dry/validation/schema/definition.rb#L13
As @zacholauson mentioned, the main issue is located at this sentence `Validation.Result(input[name], predicate.(input), self)` https://github.com/dryrb/dry-validation/blob/master/lib/dry/validation/rule/key.rb#L13  which evaluates input before predicate call. I think that we should do it after predicate call. If we do it after then we can analyse result of `key?` predicate call and only if it returns success?=true then evaluate input as `input[name]`. With current approach I can't easily implement `attr` DSL, because when object doesn't respond to specific attribute then I get `NoMethod` immediately even before `attr?` predicate returns information if object respond to this attr or not. 
Here is the temporary workaround for this https://github.com/dryrb/dry-validation/pull/29/files#diff-3cdf1f0b157c5acd91f388afa44c054cR13